util/grpc/logging.go

```diff
@@ -51,44 +54,64 @@ func (j *jsonpbMarshalleble) MarshalJSON() ([]byte, error) {
 	return b.Bytes(), nil
 }
 
-type loggingServerStream struct {
-	grpc.ServerStream
+type reporter struct {
+	ctx       context.Context
 	entry     *logrus.Entry
 	logClaims bool
 	info      string
 }
 
-func (l *loggingServerStream) SendMsg(m any) error {
-	return l.ServerStream.SendMsg(m)
-}
+func (r *reporter) PostCall(_ error, _ time.Duration) {}
+
+func (r *reporter) PostMsgSend(_ any, _ error, _ time.Duration) {}
 
-func (l *loggingServerStream) RecvMsg(m any) error {
-	err := l.ServerStream.RecvMsg(m)
+func (r *reporter) PostMsgReceive(payload any, err error, _ time.Duration) {
 	if err == nil {
-		logRequest(l.ServerStream.Context(), l.entry, l.info, m, l.logClaims)
+		logRequest(r.ctx, r.entry, r.info, payload, r.logClaims)
 	}
-	return err
 }
 
-func PayloadStreamServerInterceptor(entry *logrus.Entry, logClaims bool, decider grpc_logging.ServerPayloadLoggingDecider) grpc.StreamServerInterceptor {
-	return func(srv any, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
-		if !decider(stream.Context(), info.FullMethod, srv) {
-			return handler(srv, stream)
-		}
-		logEntry := entry.WithFields(ctx_logrus.Extract(stream.Context()).Data)
-		newStream := &loggingServerStream{ServerStream: stream, entry: logEntry, logClaims: logClaims, info: "received streaming call " + info.FullMethod}
-		return handler(srv, newStream)
+func PayloadStreamServerInterceptor(entry *logrus.Entry, logClaims bool, decider func(context.Context, interceptors.CallMeta) bool) grpc.StreamServerInterceptor {
+	return selector.StreamServerInterceptor(interceptors.StreamServerInterceptor(reportable(entry, logClaims)), selector.MatchFunc(decider))
+}
+
+func PayloadUnaryServerInterceptor(entry *logrus.Entry, logClaims bool, decider func(context.Context, interceptors.CallMeta) bool) grpc.UnaryServerInterceptor {
+	return selector.UnaryServerInterceptor(interceptors.UnaryServerInterceptor(reportable(entry, logClaims)), selector.MatchFunc(decider))
+}
+
+func reportable(entry *logrus.Entry, logClaims bool) interceptors.CommonReportableFunc {
+	return func(ctx context.Context, c interceptors.CallMeta) (interceptors.Reporter, context.Context) {
+		return &reporter{
+			ctx:       ctx,
+			entry:     entry.WithFields(ctx_logrus.Extract(ctx).Data),
+			info:      fmt.Sprintf("received %s call %s", c.Typ, c.FullMethod()),
```

Comment:
To avoid a significant change to the log structure, can we convert c.Typ to either "streaming" or "unary"? I just want to avoid breaking any logs-based systems that depend on this string that hasn't changed in a long time.